## 1 Eliminacja wartości odstających

means =

 Columns 1 through 6:

   1.8679e-01   1.4839e-02   2.1045e-01   2.0882e-01   7.9658e+01   1.0604e+00

 Column 7:

   9.0846e-03

ans = min 0.125 for class 1 is at 642
ans = min 0 for class 2 is at 642
ans = min 0 for class 3 is at 642
ans = min 0 for class 4 is at 642
ans = min -5.4538e-09 for class 5 is at 393
ans = min -1.214e-07 for class 6 is at 25
ans = min -3.26424e-09 for class 7 is at 539
ans = max 5.12127 for class 1 is at 186
ans = max 25.8229 for class 2 is at 186
ans = max 382.139 for class 3 is at 186
ans = max 380.858 for class 4 is at 186
ans = max 145297 for class 5 is at 186
ans = max 1934.12 for class 6 is at 186
ans = max 16.5702 for class 7 is at 186
means =

 Columns 1 through 6:

   1.8404e-01   6.9109e-04   9.4984e-04   1.5044e-05  -7.5376e-10   2.9695e-07

 Column 7:

  -9.6408e-11

ans = min 0.148148 for class 1 is at 186min 642 for class 
ans = min 3.51422e-09 for class 2 is at 704
ans = min 7.53088e-11 for class 3 is at 816
ans = min 5.13337e-11 for

Usuwam ze zbioru uczącego próbki 642 (wartości skrajnie małe dla większości klas) oraz 186 (wartości skrajnie duże dla większości klas). Sensowność usinięcia tych dwóch próbek widoczna jest skali wykresów generowanych przez plot2features. Przed eliminacją dane tworzyły dwa zgrupowania – punktów odstających i reszty. Po usunięciu wartości odstających skala wykresów pozwala zauważyć to jak wyglądają różnice między klasami.


In [ ]:
for i=1:size(Combs)
    ttemp = train;
    testtemp = test;

    first_idx = Combs(i, 1);
    second_idx = Combs(i, 2);

    ttemp = ttemp(:, [1 first_idx second_idx]);
    testtemp = testtemp(:, [1 first_idx second_idx]);

    % to nie jest najros�dniejszy wyb�r; 4 i 6 na pewno trzeba zmieni�
    % tutaj jawnie tworz� struktur� z parametrami dla klasyfikatora Bayesa
    % (po prawdzie, to dla funkcji licz�cej g�sto�� prawdobie�stwa) z za�o�eniem,
    % �e cechy s� niezale�ne

    pdfindep_para = para_indep(ttemp);
    pdfmulti_para = para_multi(ttemp);
    pdfparzen_para = para_parzen(ttemp, 0.001);
    % w sprawozdaniu trzeba podawa� szeroko�� okna (nie liczymy tego parametru z danych!)

    % wyniki do punktu 3
    base_ercf = zeros(1,3);
    base_ercf(1) = mean(bayescls(testtemp(:,2:end), @pdf_indep, pdfindep_para) != testtemp(:,1));
    base_ercf(2) = mean(bayescls(testtemp(:,2:end), @pdf_multi, pdfmulti_para) != testtemp(:,1));
    base_ercf(3) = mean(bayescls(testtemp(:,2:end), @pdf_parzen, pdfparzen_para) != testtemp(:,1));
    disp(Combs(i, :))
    disp(base_ercf)
end

  ### 2 Błędy dla kombinacji
   
   |kombinacja | pdfindep_para | pdfmulti_para | pdfparzen_para|
   |-----------|---------------|---------------|---------------|
   |2   3      |0.17818        |0.17873        |   0.21765     |
   |2   4      |0.0263158      |0.0049342      |0.0241228      |
   |2   5      |0.075658       |0.059759       |0.270833       |
   |2   6      |0.13706        |0.10197        |0.27686        |
   |2   7      |0.102522       |0.097588       |0.276864       |
   |2   8      |0.22752        |0.22862        |0.27686        |
   |3   4      |0.021382       |0.020833       |0.016996       |
   |3   5      |0.13158        |0.12774        |0.23794        |
   |3   6      |0.15570        |0.14912        |0.31469        |
   |3   7      |0.16228        |0.15406        |0.31469        |
   |3   8      |0.28125        |0.28399        |0.31469        |
   |4   5      |0.15406        |0.14583        |0.23958        |
   |4   6      |0.15844        |0.15132        |0.26864        |
   |4   7      |0.23136        |0.21217        |0.26864        |
   |4   8      |0.21272        |0.21272        |0.26864        |
   |5   6      |0.37500        |0.31305        |0.47149        |
   |5   7      |0.32511        |0.26919        |0.47149        |
   |5   8      |0.38158        |0.37829        |0.47149        |
   |6   7      |0.32675        |0.28947        |0.54331        |
   |6   8      |0.40077        |0.40241        |0.57346        |
   |7   8      |0.36842        |0.36678        |0.54331        |
   
Najbardziej korzystna jest zatem kobinacja cech 3 i 4.

Dla apriori = 0.25
  |kombinacja | pdfindep_para | pdfmulti_para | pdfparzen_para|
  |-----------|---------------|---------------|---------------|
  | 3  4      | 0.021382      | 0.020833      |0.016996       |
  | 4  6 | 0.158443  | 0.151316 |0.268640| 
  | 3  4 |0.021382|0.020833|0.016996|

## 3 

|cechy|pdfindep_para|pdfmulti_para|pdfparzen_para|
|--------|
|3  4|0.021382|0.020833|0.016996|

## 4 redukcja zbioru trenującego
|czesc|powtorzenie|cechy|pdfindep_para|pdfmulti_para|pdfparzen_para|
|----------|
|0.100000|1|3  4|0.020285|0.019189|0.047149|
|0.100000|2|3  4|0.020285|0.019189|0.047149|
|0.100000|3|3  4|0.020285|0.019189|0.047149|
|0.100000|4|3  4|0.020285|0.019189|0.047149|
|0.100000|5|3  4|0.020285|0.019189|0.047149|
|0.250000|1|3  4|0.019189|0.019189|0.044956|
|0.250000|2|3  4|0.019189|0.019189|0.044956|
|0.250000|3|3  4|0.019189|0.019189|0.044956|
|0.250000|4|3  4|0.019189|0.019189|0.044956|
|0.250000|5|3  4|0.019189|0.019189|0.044956|
|0.500000|1|3  4|0.021930|0.021930|0.037829|
|0.500000|2|3  4|0.021930|0.021930|0.037829|
|0.500000|3|3  4|0.021930|0.021930|0.037829|
|0.500000|4|3  4|0.021930|0.021930|0.037829|
|0.500000|5|3  4|0.021930|0.021930|0.037829|

## 5 różne szerokości okna Parzen

|parzen width|error|
|-------------
|0.000100|0.029605|
|0.000500|0.028509|
|0.001000|0.037829|
|0.005000|0.056469|
|0.010000|0.052632|

## 6 redukcja zbioru testującego
|czesc|powtorzenie|cechy|pdfindep_para|pdfmulti_para|pdfparzen_para|
|----------|
|0.500000|1|3  4|0.013889|0.012427|0.029240|
|0.500000|2|3  4|0.013889|0.012427|0.029240|
|0.500000|3|3  4|0.013889|0.012427|0.029240|
|0.500000|4|3  4|0.013889|0.012427|0.029240|
|0.500000|5|3  4|0.013889|0.012427|0.029240|
